### ISSUE TITLE
Templatization of SQOperator transformations

### ISSUE THREAD

**Issue by lelemmen** [GQCG/GQCP#417]
**Describe the current state of a part of the code**
Right now, we can only transform `SQOperators` with `SquareMatrices` if they have the same scalar representation.

**Describe a change that would improve the code base**
It makes sense to allow the in-place transformation of `SQOperators<OperatorScalar>` with `TransformationMatrix<TransformationScalar>` if `OperatorScalar * TransformationScalar` is again an `OperatorScalar`. For non-inplace transformations, the return type should then become `OperatorScalar * TransformationScalar`.

**Additional context**
Since `TransformationMatrix` hasn't been implemented yet, it would be best to postpone the resolution of this issue to after #409.

**Comment by lelemmen** [comment]
After having introduced `BasisTransformable`, this isn't on the menu anymore.
